src/ui/public/vis/lib/calculate_object_hash.js

```diff
@@ -0,0 +1,60 @@
+function pad(hash, len) {
+  while (hash.length < len) {
+    hash = '0' + hash;
+  }
+  return hash;
+}
+
+function fold(hash, text) {
+  let i;
+  let chr;
+  let len;
+  if (text.length === 0) {
+    return hash;
+  }
+  for (i = 0, len = text.length; i < len; i++) {
+    chr = text.charCodeAt(i);
+    hash = ((hash << 5) - hash) + chr;
+    hash |= 0;
+  }
+  return hash < 0 ? hash * -2 : hash;
+}
+
+function foldObject(hash, o, seen) {
+  return Object.keys(o).sort().reduce(foldKey, hash);
```

Comment:
this is more of a generic comment, but I think it applies here.

Functional style for this type of code, with deep nested loops, is incredibly crufty, especially in a language like Javascript. This spams the heap, in this case linearly with the number of nested objects, and all will need to be garbage collected. That in turn causes hickups for each sweep, especially noticeable in visualiztions where you want high interactivity. Unlike a "real" functional language, JS doesn't optimize for this at all.

Not saying we have to remove it now, since this PR is already an improvement over the other, but we do this type of stuff a lot in Kibana, and when we're talking about mem-issues, frequent allocation of objects is a real concern.

Reply:
i don't understand what you mean, but some arguments in favor of this:
- stringify would have to do pretty much the same ?
- deepEquals would have to do pretty much the same ?
- this is only run once when rendering visualization, so doesn't block the interactivity
- memory footprint after this change is largely improved, instead of 3GB with pie chart with 5 levels it uses 100MB, and is cleaned up correctly on new renders (which was not the case with stringify for some reason)

Reply:
I mean that for efficient JS-code, object allocation is a concern by definition. Basically, anytime we do `new`, `funcion`, `()=>`, `{}`, or `[]`. If we create a lot of objects in tight loops (as is the case here), the GC will eventually clean this up and hickup, and you have no control over when this will happen (e.g. perhaps while people are panning a map). So this is mainly a concern for apps that have code which executes frequently and in application which require high-interactivity (visualizations, video-games, maps, ...).

These are micro-optimizations of course, but they add up in the long run. It's something to keep in mind overall imo, especially for visualization-heavy UXs like Kibana.